pkg/core.go

```diff
@@ -1,6 +1,50 @@
 package datago
 
-import "context"
+import (
+	"context"
+	"sync/atomic"
+)
+
+type BufferedChan[T any] struct {
+	_channel      chan T
+	current_items int32
+	max_items     int32
+	open          bool
+}
+
+// --- Simple buffered channel implementation ---------------------------------------------------------------------------------------------------------------------------------------------------------------
+// Make it possible to track the current channel status from the outside
+
+func NewBufferedChan[T any](max_items int32) BufferedChan[T] {
+	return BufferedChan[T]{_channel: make(chan T, max_items), current_items: 0, max_items: max_items, open: true}
+}
+
+func (b *BufferedChan[T]) Send(item T) {
+	b._channel <- item
+
+	// Small perf hit, not sure it's worth it
+	atomic.AddInt32(&b.current_items, 1)
```

Comment:
concurrent accesses so this needs to be atomic, not too nice perf wise (probably why this is not in the std) but at our sampling serving speeds it's not an actual issue